batch/batch/cloud/azure/resource_utils.py

```diff
@@ -0,0 +1,135 @@
+
+log = logging.getLogger('resource_utils')
+
+# https://docs.microsoft.com/en-us/azure/virtual-machines/vm-naming-conventions
+MACHINE_TYPE_REGEX = re.compile(
+    r'(?P<typ>[^_]+)_(?P<family>[A-Z])(?P<sub_family>[^\d])?(?P<cpu>\d+)(-(?P<constrained_cpu>\d+))?(?P<additive_features>[^_]+)?(_((?P<accelerator_type>[^_]+)_)?(?P<version>.*)?)?'
+)
+
+AZURE_MAX_PERSISTENT_SSD_SIZE_GIB = 32 * 1024
+
+
+azure_local_ssd_size_per_core_by_worker_type = {
+    'D': 37.5,
+    'E': 37.5,
+    'F': 8,
+}
+
+
+azure_valid_cores_from_worker_type = {
+    'D': [2, 4, 8, 16, 32, 48, 64],
+    'E': [2, 4, 8, 16, 20, 32, 48, 64],
+    'F': [2, 4, 8, 16, 32, 48, 64, 72],
+}
+
+
+azure_valid_machine_types = []
+for cores in azure_valid_cores_from_worker_type['D']:
+    azure_valid_machine_types.append(f'Standard_D{cores}d_v4')
+    azure_valid_machine_types.append(f'Standard_D{cores}s_v4')
+for cores in azure_valid_cores_from_worker_type['E']:
+    azure_valid_machine_types.append(f'Standard_E{cores}d_v4')
+    azure_valid_machine_types.append(f'Standard_E{cores}s_v4')
+for cores in azure_valid_cores_from_worker_type['F']:
+    azure_valid_machine_types.append(f'Standard_F{cores}s_v2')
+
+
+azure_memory_to_worker_type = {
+    'lowmem': 'F',
+    'standard': 'D',
+    'highmem': 'E'
+}
+
+
+class MachineTypeParts:
+    @staticmethod
+    def from_dict(data: dict) -> 'MachineTypeParts':
+        constrained_cpu = data['constrained_cpu']
+        if constrained_cpu is not None:
+            constrained_cpu = int(constrained_cpu)
+        return MachineTypeParts(data['typ'], data['family'], data['sub_family'],
+                                int(data['cpu']), constrained_cpu,
+                                data['additive_features'], data['accelerator_type'],
+                                data['version'])
+
+    def __init__(self, typ: str, family: str, sub_family: Optional[str], cores: int,
+                 constrained_cpu: Optional[int], additive_features: Optional[str],
+                 accelerator_type: Optional[str], version: Optional[str]):
+        self.typ = typ
+        self.family = family
+        self.sub_family = sub_family
+        self.cores = cores
+        self.constrained_cpu = constrained_cpu
+        self.additive_features = additive_features
+        self.accelerator_type = accelerator_type
+        self.version = version
+
+
+def azure_machine_type_to_parts(machine_type: str) -> Optional[MachineTypeParts]:
+    match = MACHINE_TYPE_REGEX.fullmatch(machine_type)
+    if match is None:
+        return match
+    return MachineTypeParts.from_dict(match.groupdict())
+
+
+def azure_machine_type_to_worker_type_and_cores(machine_type: str) -> Tuple[str, int]:
+    maybe_machine_type_parts = azure_machine_type_to_parts(machine_type)
+    if maybe_machine_type_parts is None:
+        raise ValueError(f'bad machine_type: {machine_type}')
+    # local_ssd = maybe_machine_type_parts.additive_features == 'd' or maybe_machine_type_parts.family == 'F'
```

Comment:
I ended up not needing to return the local SSD, but it might be good to add it as a return type in order to assert that the machine type has the right option for the local ssd when creating instances, etc. in Azure.